src/middlewared/middlewared/plugins/sysdataset.py

```diff
@@ -480,76 +607,93 @@ def __get_datasets(self, pool, uuid):
         ]
 
     @private
-    async def migrate(self, _from, _to):
-
-        config = await self.config()
+    def migrate(self, _from, _to):
+        """
+        Migrate system dataset to a new pool. If it is moving from
+        an existing pool, then the new datasets are mounted in
+        the middleware rundir temprorarily so that data can be
+        rsynced from the old pool.
+        """
+        config = self.middleware.call_sync('systemdataset.config')
 
-        await self.__setup_datasets(_to, config['uuid'])
+        self.middleware.call_sync('systemdataset.setup_datasets', _to, config['uuid'])
 
         if _from:
-            path = '/tmp/system.new'
-            if not os.path.exists('/tmp/system.new'):
-                os.mkdir('/tmp/system.new')
+            path = f'{MIDDLEWARE_RUN_DIR}/system.new'
+            if not os.path.exists(f'{MIDDLEWARE_RUN_DIR}/system.new'):
```

Comment:
os.mkdir(exist_ok=True)?

Reply:
That's for os.makedirs() rather than os.mkdir(). I wrapped in contextlib.suppress though.